feat: remove Revert.dev dependency from Pipedrive integration

anikdhabal opened the pull request:
## What does this PR do?

Removes the dependency on Revert.dev from the Pipedrive CRM integration and implements direct OAuth2 authentication with Pipedrive's native API. The changes follow the same patterns used by other CRM integrations in the codebase (particularly HubSpot and Zoom).

- Fixes #16797

## Summary

**Key Changes:**
- **OAuth Flow**: Replaced Revert redirect with direct Pipedrive OAuth2 authorization
- **API Integration**: Converted all API calls from Revert's unified format to Pipedrive's native REST API
- **Token Management**: Implemented proper OAuth token refresh using Cal.com's `OAuthManager` pattern (same as Zoom, Google Calendar, etc.)
- **Environment Variables**: Removed `REVERT_*` variables, now requires `PIPEDRIVE_CLIENT_ID` and `PIPEDRIVE_CLIENT_SECRET`
- **Data Mapping**: Mapped contact and activity data between Revert's unified format and Pipedrive's native structure

**Breaking Change**: Existing Pipedrive integrations will need to be reconfigured with new OAuth credentials.

### Updates since last revision

- Refactored `CrmService.ts` to use Cal.com's `OAuthManager` for proper token management
- Added `getPipedriveAppKeys` helper function for client credentials extraction with Zod validation
- Implemented `requestPipedrive` helper method that uses `OAuthManager.request()` for all API calls
- Added proper token refresh callbacks (`isTokenObjectUnusable`, `isAccessTokenUnusable`) to detect Pipedrive error responses
- Added validation for token endpoint response in callback.ts - now returns proper error when OAuth code exchange fails
- Merged latest main branch to ensure compatibility

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TB
    subgraph "OAuth Flow"
        A["api/add.ts<br/>OAuth Authorization"]:::major-edit
        B["api/callback.ts<br/>Token Exchange"]:::major-edit
    end
    
    subgraph "CRM Integration"
        C["lib/CrmService.ts<br/>Pipedrive API Client"]:::major-edit
        D["lib/getPipedriveAppKeys.ts<br/>Credentials Helper"]:::major-edit
        E["config.json<br/>App Configuration"]:::minor-edit
    end
    
    subgraph "External APIs"
        F["Pipedrive OAuth API<br/>oauth.pipedrive.com"]:::context
        G["Pipedrive REST API<br/>api.pipedrive.com"]:::context
    end
    
    A --> B
    B --> C
    C --> G
    A --> F
    B --> F
    C --> D
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#ADD8E6
    classDef context fill:#FFFFFF
```

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. N/A - environment variable changes are documented in .env.appStore.example
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

**Environment Variables Required:**
- `PIPEDRIVE_CLIENT_ID` - OAuth client ID from Pipedrive developer portal
- `PIPEDRIVE_CLIENT_SECRET` - OAuth client secret from Pipedrive developer portal

**Test Plan:**
1. Set up a Pipedrive developer account and create an OAuth app
2. Configure the environment variables above
3. Test the integration setup flow in Cal.com (OAuth authorization)
4. Create a test booking and verify it creates contacts and activities in Pipedrive
5. Test updating and canceling bookings
6. **Important**: Test token refresh by waiting for token expiry or manually expiring the token

**Review Checklist for Human:**
- [ ] Verify OAuthManager integration correctly handles token refresh
- [ ] Test `isTokenObjectUnusable` callback detects "invalid_grant" errors
- [ ] Test `isAccessTokenUnusable` callback detects 401 responses
- [ ] Verify `api_domain` is correctly preserved and updated in token object
- [ ] Test contact creation and search with real Pipedrive data
- [ ] Test activity/meeting CRUD operations

## Notes

- **OAuth Scopes**: Using comprehensive scopes: `deals:read,deals:write,persons:read,persons:write,activities:read,activities:write`
- **API Mapping**: Activities are used for meetings/events instead of a separate events API, following Pipedrive's data model
- **Token Storage**: Uses Cal.com's standard credential storage with `expiry_date` in milliseconds
- **No automated tests**: This integration requires real Pipedrive API credentials to test properly

---

**Session Info:** 
- Link to Devin run: https://app.devin.ai/sessions/e175c40829e243e9a355616269b10acf
- Originally requested by: anik@cal.com (@anikdhabal)

devin-ai-integration[bot] committed:
feat: remove Revert.dev dependency from Pipedrive integration

- Replace Revert API calls with direct Pipedrive OAuth2 and REST API
- Update OAuth flow to use Pipedrive's authorization endpoint
- Implement token exchange in callback handler
- Map Revert's unified API format to Pipedrive's native API structure
- Remove REVERT_* environment variables from configuration
- Update app config to reflect direct Pipedrive integration
- Follow patterns from other CRM integrations like HubSpot

Breaking change: Requires PIPEDRIVE_CLIENT_ID and PIPEDRIVE_CLIENT_SECRET
environment variables instead of REVERT_* variables

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

devin-ai-integration[bot] committed:
feat: implement proper token refresh logic for Pipedrive OAuth

- Add token refresh implementation following Pipedrive OAuth2 specification
- Use proper Basic Auth for token refresh requests
- Handle token refresh errors with appropriate logging
- Update access token and expiry date after successful refresh
- Add note about database credential update requirement

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

devin-ai-integration[bot] committed:
feat: add database credential updates for Pipedrive token refresh

- Import updateTokenObjectInDb utility for proper credential management
- Store credentialId in constructor for database updates
- Update token refresh logic to persist refreshed tokens to database
- Follow Cal.com patterns for OAuth credential management
- Ensure tokens are properly updated after successful refresh

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

hariombalhara reviewed (changes requested):
@anikdhabal We have OAuthManager that we use for Google Calendar and Microsoft Office. I believe zoom too.
Can you check if we can use that here.

The benefit with OAuthManager approach is that it automatically handles invalidation of credential , plus supports credential sync feature out of the box.

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

devin-ai-integration[bot] committed:
refactor: use OAuthManager for Pipedrive token management

- Refactor CrmService.ts to use OAuthManager instead of manual token refresh
- Create getPipedriveAppKeys helper function for client credentials
- Update callback.ts to use the new helper function
- Add requestPipedrive helper method for all API calls
- Implement proper token refresh callbacks (isTokenObjectUnusable, isAccessTokenUnusable)

This addresses the review comments from hariom and cubic on PR #22492.

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

devin-ai-integration[bot] committed:
Merge remote-tracking branch 'origin/main' into devin/remove-revert-pipedrive-1752520608

devin-ai-integration[bot] committed:
fix: validate token endpoint response before storing Pipedrive credentials

Co-Authored-By: unknown <>

hariombalhara reviewed (dismissed):
Code LGTM now !!

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608

anikdhabal committed:
Add credential property to PipedriveCrmService

anikdhabal committed:
Merge branch 'main' into devin/remove-revert-pipedrive-1752520608
